### Definição algoritmo UPMGA

In [18]:
import pandas as pd
matrix = pd.read_csv('matriz_distancias.csv', sep = '\t', index_col= 'index')

def encontraMenorDistancia(matrix):
  min = matrix.iloc[0,-1]
  for i in matrix.iterrows():
    for j in matrix:
      if (i[1][j] <= min and i[1][j] > 0):
        min = i[1][j]
        gen1 = i[0]
        gen2 = j
  return min, (gen1, gen2)

def substituiPar(matrix, par):
    print(matrix)
    matrix[str(par).replace("'",'')] = matrix.apply(lambda x: (x[par[0]] + x[par[1]])/2)
    matrix.drop(list(par), axis =1, inplace=True)
    matrix.drop(list(par), inplace = True)
    vec = list(matrix[str(par).replace("'",'')])
    vec.append(0)
    matrix.loc[str(par).replace("'",'')] = vec
    print(min)
    

def upmga(matrix):
  while len(matrix.columns) > 1:
    min, par = encontraMenorDistancia(matrix)
    substituiPar(matrix, par)
  return (matrix.columns[0])
  

### Questão 1

In [19]:
matrix = pd.read_csv('matriz_distancias.csv', sep = '\t', index_col= 'index')
upmga(matrix)


              A         B         C         D         E         F
index                                                            
A      0.000000  1.423067  1.246474  2.482968  1.602307  0.744842
B      1.423067  0.000000  1.898534  3.135029  2.254368  1.943034
C      1.246474  1.898534  0.000000  1.482011  0.601351  1.766440
D      2.482968  3.135029  1.482011  0.000000  1.267805  3.002934
E      1.602307  2.254368  0.601351  1.267805  0.000000  2.122274
F      0.744842  1.943034  1.766440  3.002934  2.122274  0.000000
2.4097011
               A         B         D         F    (E, C)
index                                                   
A       0.000000  1.423067  2.482968  0.744842  1.424390
B       1.423067  0.000000  3.135029  1.943034  2.076451
D       2.482968  3.135029  0.000000  3.002934  1.374908
F       0.744842  1.943034  3.002934  0.000000  1.944357
(E, C)  1.424390  2.076451  1.374908  1.944357  0.000000
2.4097011
               B         D    (E, C)    (F, A)
index 

'(((F, A), B), ((E, C), D))'

### Questão 2 - A

|Sequencia|Organismo|Max Score|
|---|:---:|---:|
|seq_1|Bacillus anthracis|18465|
|seq_2|Clostridium tetani|18465|
|seq_3|Escherichia coli|18222|
|seq_4|Mycobacterium tuberculosis|18465|
|seq_5|Pseudomonas aerugin|18465|
|seq_6|Salmonella enteric|18465|
|seq_7|Staphylococcus aureus|18465|
|seq_8|Streptococcus pneumoni|18465|
|seq_9|Treponema pallidum|18465|
|seq_10|Vibrio cholerae|18465|

### Questão 2 - B

In [37]:
import sys

gap = -2
mismatch = -1
match = 1



def preencheMatrix(matrix):
  for i in range(1, len(matrix[0])):
    matrix[1].append((i-1)*gap)  
  for j in range(2, len(matrix)):
    matrix[j].append((j-1)*gap)
  for x in range(2, len(matrix[0])):
    for i in range(2, len(matrix)):
      if matrix[i][0] == matrix[0][x]:
          matrix[i].append(max(matrix[i-1][x-1] +match, matrix[i-1][x] + gap, matrix[i][x-1] + gap))
      else:
          matrix[i].append(max(matrix[i-1][x-1] + mismatch, matrix[i-1][x] +gap, matrix[i][x-1] +gap))


def calculaAlinhamento(matrix):
  seq1 = ""
  seq2 = ""
  i = len(matrix)-1
  j = len(matrix[0])-1
  score = matrix[i][j]
  while i > 1 or j > 1:
    if j <= 1:
      seq1 = '-' + seq1
      seq2 = matrix[i][0] + seq2
      i = i-1
    elif i <= 1:
      seq1 = matrix[0][j] + seq1
      seq2 = '-' + seq2
      j = j-1
    elif matrix[0][j] == matrix[i][0]:
      seq1 = matrix[0][j] + seq1
      seq2 = matrix[i][0] + seq2
      i = i-1
      j = j-1
    else:
      maximum = max(matrix[i-1][j-1], matrix[i][j-1], matrix[i-1][j])
      if maximum == matrix[i-1][j-1]:
        seq1 = matrix[0][j] + seq1
        seq2 = matrix[i][0] + seq2
        i = i-1
        j = j-1
      elif maximum == matrix[i-1][j]:
        seq1 = '-' + seq1
        seq2 = matrix[i][0] + seq2
        i = i-1
      elif maximum == matrix[i][j-1]:
        seq1 = matrix[0][j] + seq1
        seq2 = '-' + seq2
        j = j-1
  return score, seq1, seq2

def calculaScore(seq1, seq2):
  i = len(seq1)
  j = len(seq2)
  linha_anterior = list(range(0,len(seq1)*gap, gap))
  print(linha_anterior)
    


seq1 = open('seq1.fna', 'r')
seq1 = (seq1.read().replace('\n','')[:100])
print(len(seq1))
seq2 = open('seq9.fna', 'r')
seq2 = (seq2.read().replace('\n','')[:100])


print(seq1)
print(seq2)
matrix = montaMatrix(seq1, seq2)
preencheMatrix(matrix)
result = calculaScore(matrix)

print('Score: ',result)



100
ATATTTTTTCTTGTTTTTTATATCCACAAACTCTTTTCGTACTTTTACACAGTATATCGTGTTGTGGACAATTTTATTCCACAAGGTATTGATTTTGTGG
TAGATGGACGCAGTAGGGTATGAAGTATTCTGGAACGAGACACTCAGCCAGATACGGAGTGAATCGACCGAAGCAGAATTTAACATGTGGTTTGCTCATT
Score:  -25


In [45]:
import glob
files = glob.glob("s*")
for i in range (7,11):
    for j in range(1,11):
        if i < j:
            seq_a = open(f'seq{i}.fna', 'r').read()
            seq_b = open(f'seq{j}.fna', 'r').read()
            matrix = montaMatrix(seq_a, seq_b)
            preencheMatrix(matrix)
            result = calculaScore(matrix)
            print(f'seq_{i} vs seq_{j}: {result}')
    

seq_7 vs seq_8: -752
seq_7 vs seq_9: -1179
seq_7 vs seq_10: -1280
seq_8 vs seq_9: -1236
seq_8 vs seq_10: -1332
seq_9 vs seq_10: -1295


seq_1 vs seq_2: -1109
seq_1 vs seq_3: -1388
seq_1 vs seq_4: -1625
seq_1 vs seq_5: -1522
seq_1 vs seq_6: -1528
seq_1 vs seq_7: 1617
seq_1 vs seq_8: -518
seq_1 vs seq_9: -1213
seq_1 vs seq_10: -1297
seq_2 vs seq_3: -1541
seq_2 vs seq_4: -1958
seq_2 vs seq_5: -1937
seq_2 vs seq_6: -1717
seq_2 vs seq_7: -1032
seq_2 vs seq_8: -1194
seq_2 vs seq_9: -1329
seq_2 vs seq_10: -1556
seq_3 vs seq_4: -1262
seq_3 vs seq_5: -1220
seq_3 vs seq_6: -1167
seq_3 vs seq_7: -1450
seq_3 vs seq_8: -1448
seq_3 vs seq_9: -1357
seq_3 vs seq_10: -1240
seq_4 vs seq_5: -749
seq_4 vs seq_6: -1057
seq_4 vs seq_7: -1303
seq_4 vs seq_8: -1712
seq_4 vs seq_9: -1394
seq_4 vs seq_10: -1289
seq_5 vs seq_6: -1040
seq_5 vs seq_7: -1780
seq_5 vs seq_8: -1972
seq_5 vs seq_9: -1454
seq_5 vs seq_10: -1324
seq_6 vs seq_7: -1552
seq_6 vs seq_8: -1605
seq_6 vs seq_9: -1296
seq_6 vs seq_10: -1255
seq_7 vs seq_8: -752
seq_7 vs seq_9: -1179
seq_7 vs seq_10: -1280
seq_8 vs seq_9: -1236
seq_8 vs seq_10: -1332
seq_9 vs seq_10: -1295

In [158]:
gap = -2
mismatch = -1
match = 1

def calculaScore(seq1, seq2):
  i = 1
  j = 0
  seq1 = ' '+ seq1
  seq2 = ' ' + seq2
  linha_anterior = list(range(0,(len(seq1))*gap, gap))
  while i < len(seq1):
    atual = [(i)*gap]
    for j in range(1,len(seq2)):
      if seq1[i] == seq2[j]:
        atual.append(max(linha_anterior[j -1]+match, linha_anterior[j] + gap, atual[-1] + gap))
      else:
        atual.append(max(linha_anterior[j -1]+mismatch, linha_anterior[j] + gap, atual[-1] + gap))
    linha_anterior = atual
    i += 1
  print(atual[-1])


seq1 = open('seq1.fna', 'r')
seq1 = (seq1.read().replace('\n','')[3:200])
seq2 = open('seq2.fna', 'r')
seq2 = (seq2.read().replace('\n','')[3:200])
print(seq1)
print(seq2)
calculaScore(seq1, seq2)

TTTTTTCTTGTTTTTTATATCCACAAACTCTTTTCGTACTTTTACACAGTATATCGTGTTGTGGACAATTTTATTCCACAAGGTATTGATTTTGTGGATAACTTTCTTAATTTCATTGCTATAGCTACTTTTTTTTGATATTATAGTTGTGTTTTCACTTTGAATAAGTTTTCCACATCTTTATCTTATCCACAATT
TTTCTTTGAACTTGTCAACACTTTTTTAAAAAATGTTTTTATGTTTGAAATAACTTGTAAAAATTTATGTTGTTCCTTTGCAACAACGTATTTTATTCTAACACAATTTAATTTAACTATTGAGTATTAAAACCTCATTTAATGCTATTATCTTTTATTTATTTATTATTTTTAGATTTCTCTCCATATTCCACATA
-3
